### PR TITLE
Thrift: stats for downstream connection close

### PR DESCRIPTION
**Pull request by JuniorHsu**
Signed-off-by: kuochunghsu <kuochunghsu@pinterest.com>

Commit Message:
We want to keep downstream connections as much as possible since SR drops usually come from cx close.
Therefore, we want more stats for reasoning.

Moreover, adding more debug log to see reset reason.

Unfortunately we didn't test for partial response during connection close, which might be an followup.

Additional Description: 
Risk Level: low
Testing: unit
Docs Changes: current.yaml
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by zuercher** [comment]
`Code coverage for source/extensions/filters/network/thrift_proxy/router is lower than limit of 96.4 (96.1)`

I expect you need to exercise the path for `downstream_cx_partial_request_close_` in tests to get the coverage up.

**Comment by JuniorHsu** [comment]
> `Code coverage for source/extensions/filters/network/thrift_proxy/router is lower than limit of 96.4 (96.1)`
> 
> I expect you need to exercise the path for `downstream_cx_partial_request_close_` in tests to get the coverage up.

I add a test. Hopefully it passes the threshold. Thanks.

File: source/extensions/filters/network/thrift_proxy/router/upstream_request.cc [hunk 1]
```diff
@@ -321,6 +322,13 @@ bool UpstreamRequest::onResetStream(ConnectionPool::PoolFailureReason reason) {
     // downstream.
     if (response_underflow_ || response_state_ == ResponseState::Started) {
       ENVOY_LOG(debug, "reset downstream connection for a partial or underflow response");
+      if (response_underflow_) {
+        std::cout << "XXX response underflow" << std::endl;
```
Comment (rgs1) [inline source/extensions/filters/network/thrift_proxy/router/upstream_request.cc:326]: drop the cout

File: source/extensions/filters/network/thrift_proxy/router/upstream_request.cc [hunk 2]
```diff
@@ -321,6 +322,13 @@ bool UpstreamRequest::onResetStream(ConnectionPool::PoolFailureReason reason) {
...
+        stats_.incCloseUnderflowResponse(parent_.cluster());
+      } else {
+        std::cout << "XXX partial response" << std::endl;
```
Comment (rgs1) [inline source/extensions/filters/network/thrift_proxy/router/upstream_request.cc:329]: drop the cout (or make it a log statement)

File: test/extensions/filters/network/thrift_proxy/router_test.cc [hunk 3]
```diff
@@ -548,7 +548,7 @@ class ThriftRouterTestBase {
   }
 
   void returnResponse(MessageType msg_type = MessageType::Reply, bool is_success = true,
-                      bool is_drain = false) {
+                      bool is_drain = false, bool is_partial_response = false) {
```
Comment (rgs1) [inline test/extensions/filters/network/thrift_proxy/router_test.cc:551]: nit: `is_partial` is probably enough since we know we are talking about responses

**Review by rgs1 — APPROVED** [review]
Thanks!

**Comment by alyssawilk** [comment]
looks like this needs a main merge?

**Comment by JuniorHsu** [comment]
Did an upstream merge to resolve the conflict

**Review by zuercher — APPROVED** [review]

**Pull request merged**